A bunch of Windows pen-related improvements

Namaneo opened the pull request:
* Fixes an issue where tablet control overrides were being ignored because of the context re-creation. 
* Added `MTY_PEN_FLAG_DOUBLE_CLICK` to let the the library user interpret this event as he wish.
* Use `GetCursorPos` instead of the data reported by Wintab to ensure cursor position correctness.
* Convert `MTY_EVENT_PEN` to `MTY_EVENT_MOTION` and `MTY_EVENT_BUTTON` instead of using the auto-generated ones. This guarantees a better reactivity compared to what WinInk exposes (e.g. `DOWN` and `UP` event were both sent at the same time after the pen leaves the surface).
* Update cursor even when the window is not focus, and prevent some mechanisms from being executed during pen events (as those caused issues, probably because they are aiming real mouse events).
* If Wintab is in use, grab the focus when hovering the window without clicking. This is required because Wintab only reports data to focuses windows.

Namaneo committed:
Fix tablet controls override being ignored

Namaneo committed:
Filter out Wintab message when pen is disabled

Namaneo committed:
Add Wintab-specific pen flags

Namaneo committed:
Update MTY_PenFlag

Namaneo committed:
Better pen double-click handling through a flag

Namaneo committed:
Update spacing

Namaneo committed:
Fix range issue on MTY_WINTAB_TYPE_*

megadrago88 committed:
merge `master` into `wintab_fix-override`

megadrago88 committed:
fix irrelevant build error

RonaldH commented on src/windows/wintab.c:
```diff
@@ -272,19 +274,27 @@ void wintab_on_packetext(struct wintab *ctx, MTY_Event *evt, const PACKETEXT *pk
 	const SLIDERDATA *curr_ring = &pktext->pkTouchRing;
 	const SLIDERDATA *prev_ring = &ctx->prev_pktext.pkTouchRing;
 	if (curr_ring->nPosition != prev_ring->nPosition || curr_ring->nMode != prev_ring->nMode) {
+		uint16_t position = (uint16_t) curr_ring->nPosition;
+		if (position >= 73) // If outside of the logical 0-72 range, offset it back for event consistency
+			position -= UINT8_MAX - 73;
+
```
`curr_ring->nPosition` is 32 bits, so the value could be 65535,
Thus position is larger than 73, then position -= 255 - 73, which means that the position is still larger than 73. Do we just need to set position to 73? do we need to see if the position is larger than 255?

Namaneo replied:
We ended up with this after extensive testing of the extension, where we've seen this behavior:
* `nPosition` will range from 0 to 72 here, and can't be greater than that in a normal scenario (I think Jason can confirm)
* In case custom settings for `parsecd` are configured in the Wacom settings, the position is "inverted" within a byte range, so it ranges from 182 to 255 (no clue why, and this is not documented anywhere)

An option to make this check safer would be to set the value to zero if the position is out of the known ranges. That way, the driver's behavior is known. Would that fit?

RonaldH replied:
Double check if larger than 255 then, if larger than 255, make it 0.

Namaneo committed:
Cleanup

Namaneo committed:
Woops

Namaneo committed:
Fix locked cursor after leaving the window area

Namaneo committed:
Fix Wacom mouse mode being broken

Quick details about this one: even using mouse mode, Wintab report absolute tablet coordinates. This results in the visible cursor being constently off compared to what Wintab reports. Using GetCursorPos instead guarantees cursor position to be correct.

Namaneo committed:
Fix mouse being locked after pen has been used

Namaneo committed:
Add context overlap Wintab function

Namaneo committed:
Fix Wacom mouse mode (again)

Namaneo committed:
Fix pen/mouse switch while focused and un-focused

Namaneo committed:
Fix broken mouse mode (again again)

Namaneo committed:
Convert pen message to mouse message

Commit:
D3D9 shader compatibility with 4_0_level_9_3

Commit:
Preserve maximized state when hiding window

Commit:
Semantic adjustments, no longer need MTY_App for default frame

Commit:
When getting placement, make sure AdjustWindowRect takes into account the scale of the correct monitor

Commit:
User DPI aware AdjustWindowRect

Commit:
Default frame adjustments, wsize math

Commit:
Better restoration of frame after DPI change

Commit:
Fancy math

Namaneo committed:
Removed obsolete pieces of code

Commit:
Minor math simplification

Namaneo committed:
Grab focus on mouse move only when using a pen

Namaneo committed:
Forgot parentheses

Namaneo committed:
Small cleanup

Commit:
macOS position prep

Commit:
macOS fixes -- fullscreen behavior is not robust

Namaneo committed:
Reset pen_in_range when the window is un-focused

Commit:
macOS maximized / fullscreen normal frame behavior

Namaneo committed:
Woops

Commit:
macOS hold off on going fullscreen

Commit:
Linux progress -- normalization/denormalization sill has issues

Commit:
Cleaned up encapsulation

Commit:
Revert "D3D9 shader compatibility with 4_0_level_9_3"

This reverts commit df18db8c06361868ab9269d999b54fad6f61f711.

Commit:
Missing header

Commit:
Documentation cleanup

Namaneo committed:
Fix eraser being locked in some situations

Namaneo committed:
Convert pen to mouse using Wintab events

Commit:
Misc cleanup

Commit:
Clearer normalize function

Commit:
Windows cleanup

Kodikuu committed:
Fix JSON last item removal (#26)

Kodikuu committed:
MTY_JSONObjGetFullString (#23)

Kodikuu committed:
Move copy fallback for WIndows (#18)

Kodikuu committed:
[Minor] Ignore motion to the same position (#17)

Commit:
Linux proper maxmimized state/tracking

Commit:
Handle fullscreen properly

Namaneo committed:
Prevent motions when detach state is full

Commit:
macOS syntax fix

Commit:
macOS fullscreen reboot

Commit:
Do not assume third party API internals will execute on the same thread even if synchronous

Namaneo committed:
Filter relative mode when the pen is leaving

Commit:
Merge branch 'pos'

- Rebooted window positioning methodology
- Introduced MTY_Frame concept for saving size and normal 'restored' position
- New fullscreen behavior on macOS that behaves better with multiple windows
- D3D12 fallback to D3D11

Commit:
Merge branch 'master' into master

Commit:
Merge pull request #27 from chrisd1100/master

Sync upstream

Namaneo committed:
Remove obsolete code

Namaneo committed:
Merge branch 'master' into wintab_fix-override

Namaneo committed:
Also convert to motion when pen not in range

Commit:
Merge upstream

Kodikuu committed:
MTY_AudioReset Windows bugfix (#15)

Namaneo committed:
Merge remote-tracking branch 'origin/master' into wintab_fix-override

Namaneo committed:
Cleanup

Namaneo committed:
Filtering button events outside the window

RonaldH committed:
Merge branch 'master' into wintab_fix-override

RonaldH committed:
Merge branch 'master' into wintab_fix-override

Namaneo committed:
Filter relative mouse when the pen leaves

Namaneo committed:
Fixed multi-window conflicts

Namaneo committed:
Remove unnecessarry SetFocus

Namaneo committed:
Fix enhanced pen off on other screens

RonaldH committed:
Merge branch 'master' into wintab_fix-override
